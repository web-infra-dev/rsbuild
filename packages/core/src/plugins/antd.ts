import { getAntdMajorVersion } from '@modern-js/utils';
import type { RsbuildTarget, DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginAntd = (): DefaultRsbuildPlugin => ({
  name: `plugin-antd`,

  setup(api) {
    api.modifyRsbuildConfig((builderConfig) => {
      builderConfig.source ??= {};

      if (
        builderConfig.source.transformImport === false ||
        builderConfig.source.transformImport?.some(
          (item) => item.libraryName === 'antd',
        )
      ) {
        return;
      }

      const antdMajorVersion = getAntdMajorVersion(api.context.rootPath);
      // antd >= v5 no longer need babel-plugin-import
      // see: https://ant.design/docs/react/migration-v5#remove-babel-plugin-import
      if (antdMajorVersion && antdMajorVersion < 5) {
        builderConfig.source ??= {};
        builderConfig.source.transformImport = [
          ...(builderConfig.source.transformImport || []),
          {
            libraryName: 'antd',
            libraryDirectory: useSSR(api.context.target) ? 'lib' : 'es',
            style: true,
          },
        ];
      }
    });
  },
});

export function useSSR(target: RsbuildTarget | RsbuildTarget[]) {
  return (Array.isArray(target) ? target : [target]).some((item) =>
    ['node', 'service-worker'].includes(item),
  );
}
