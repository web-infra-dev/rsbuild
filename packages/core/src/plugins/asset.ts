import path from 'path';
import {
  getDistPath,
  getFilename,
  chainStaticAssetRule,
} from '@rsbuild/shared';
import type { DefaultRsbuildPlugin } from '@rsbuild/shared';

export function getRegExpForExts(exts: string[]): RegExp {
  const matcher = exts
    .map((ext) => ext.trim())
    .map((ext) => (ext.startsWith('.') ? ext.slice(1) : ext))
    .join('|');

  return new RegExp(
    exts.length === 1 ? `\\.${matcher}$` : `\\.(${matcher})$`,
    'i',
  );
}

export const pluginAsset = (
  assetType: 'image' | 'media' | 'font' | 'svg',
  exts: string[],
): DefaultRsbuildPlugin => ({
  name: `plugin-${assetType}`,

  setup(api) {
    api.modifyBundlerChain((chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const regExp = getRegExpForExts(exts);
      const distDir = getDistPath(config.output, assetType);
      const filename = getFilename(config.output, assetType, isProd);

      const maxSize = config.output.dataUriLimit[assetType];

      const rule = chain.module.rule(assetType).test(regExp);

      chainStaticAssetRule({
        rule,
        maxSize,
        filename: path.posix.join(distDir, filename),
        assetType,
      });
    });
  },
});
