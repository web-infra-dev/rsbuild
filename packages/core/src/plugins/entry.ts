import { castArray, color } from '@rsbuild/shared';
import type { EntryDescription } from '@rspack/core';
import { createVirtualModule } from '../helpers';
import type { RsbuildPlugin } from '../types';

export const pluginEntry = (): RsbuildPlugin => ({
  name: 'rsbuild:entry',

  setup(api) {
    api.modifyBundlerChain(async (chain, { environment, isServer }) => {
      const config = api.getNormalizedConfig({ environment });
      const { preEntry } = config.source;
      const { entry } = api.context.environments[environment];

      const injectCoreJsEntry = config.output.polyfill === 'entry' && !isServer;

      for (const entryName of Object.keys(entry)) {
        const entryPoint = chain.entry(entryName);
        const addEntry = (item: string | EntryDescription) => {
          entryPoint.add(item);
        };

        preEntry.forEach(addEntry);

        if (injectCoreJsEntry) {
          addEntry(createVirtualModule('import "core-js";'));
        }

        castArray(entry[entryName]).forEach(addEntry);
      }
    });

    api.onBeforeCreateCompiler(({ bundlerConfigs }) => {
      if (bundlerConfigs.every((config) => !config.entry)) {
        throw new Error(
          `Could not find any entry module, please make sure that ${color.cyan(
            'src/index.(ts|js|tsx|jsx|mjs|cjs)',
          )} exists, or customize entry through the ${color.cyan(
            'source.entry',
          )} configuration.`,
        );
      }
    });
  },
});
