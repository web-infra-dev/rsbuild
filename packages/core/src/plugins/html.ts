import path, { isAbsolute } from 'path';
import {
  fse,
  color,
  isURL,
  castArray,
  getMinify,
  getDistPath,
  isFileExists,
  isPlainObject,
  isHtmlDisabled,
  removeTailSlash,
  mergeChainedOptions,
} from '@rsbuild/shared';
import type {
  HtmlConfig,
  RsbuildPluginAPI,
  NormalizedConfig,
  HTMLPluginOptions,
  HtmlInjectTagDescriptor,
} from '@rsbuild/shared';
import type { HtmlTagsPluginOptions } from '../rspack/HtmlTagsPlugin';
import type { HtmlInfo } from '../rspack/HtmlBasicPlugin';
import type { RsbuildPlugin } from '../types';

export function getTitle(entryName: string, config: NormalizedConfig) {
  return mergeChainedOptions({
    defaults: '',
    options: config.html.title,
    utils: { entryName },
    useObjectParam: true,
  });
}

export function getInject(entryName: string, config: NormalizedConfig) {
  return mergeChainedOptions({
    defaults: 'head',
    options: config.html.inject,
    utils: { entryName },
    useObjectParam: true,
  });
}

const existTemplatePath: string[] = [];

export async function getTemplate(
  entryName: string,
  config: NormalizedConfig,
  rootPath: string,
): Promise<{ templatePath: string; templateContent?: string }> {
  const DEFAULT_TEMPLATE = path.resolve(
    __dirname,
    '../../static/template.html',
  );

  const templatePath = mergeChainedOptions({
    defaults: DEFAULT_TEMPLATE,
    options: config.html.template,
    utils: { entryName },
    useObjectParam: true,
  });

  if (templatePath === DEFAULT_TEMPLATE) {
    return {
      templatePath: templatePath,
    };
  }

  const absolutePath = isAbsolute(templatePath)
    ? templatePath
    : path.resolve(rootPath, templatePath);

  if (!existTemplatePath.includes(absolutePath)) {
    // Check if custom template exists
    if (!(await isFileExists(absolutePath))) {
      throw new Error(
        `Failed to resolve HTML template, please check if the file exists: ${color.cyan(
          absolutePath,
        )}`,
      );
    }

    existTemplatePath.push(absolutePath);
  }

  const templateContent = await fse.readFile(absolutePath, 'utf-8');
  return {
    templatePath: absolutePath,
    templateContent,
  };
}

export function getFavicon(
  entryName: string,
  config: {
    html: HtmlConfig;
  },
) {
  return mergeChainedOptions({
    defaults: '',
    options: config.html.favicon,
    utils: { entryName },
    useObjectParam: true,
  });
}

export function getMetaTags(
  entryName: string,
  config: { html: HtmlConfig },
  templateContent?: string,
) {
  const metaTags = mergeChainedOptions({
    defaults: {},
    options: config.html.meta,
    utils: { entryName },
    useObjectParam: true,
  });

  // `html.meta` will add charset meta by default.
  // If the custom HTML template already contains charset meta, Rsbuild should not inject it again.
  if (templateContent && metaTags.charset) {
    const charsetRegExp = /<meta[^>]+charset=["'][^>]*>/i;
    if (charsetRegExp.test(templateContent)) {
      delete metaTags.charset;
    }
  }

  return metaTags;
}

function getTemplateParameters(
  entryName: string,
  config: NormalizedConfig,
  assetPrefix: string,
): HTMLPluginOptions['templateParameters'] {
  return (compilation, assets, assetTags, pluginOptions) => {
    const { mountId, templateParameters } = config.html;
    const defaultOptions = {
      mountId,
      entryName,
      assetPrefix,
      compilation,
      webpackConfig: compilation.options,
      htmlWebpackPlugin: {
        tags: assetTags,
        files: assets,
        options: pluginOptions,
      },
    };
    return mergeChainedOptions({
      defaults: defaultOptions,
      options: templateParameters,
      utils: { entryName },
    });
  };
}

function getChunks(entryName: string, entryValue: string | string[]) {
  const dependOn = [];

  if (isPlainObject(entryValue)) {
    // @ts-expect-error assume entry is an entry object
    dependOn.push(...entryValue.dependOn);
  }

  return [...dependOn, entryName];
}

export const applyInjectTags = (api: RsbuildPluginAPI) => {
  api.modifyBundlerChain(async (chain, { CHAIN_ID }) => {
    const config = api.getNormalizedConfig();

    const tags = castArray(config.html.tags).filter(Boolean);
    const tagsByEntries = config.html.tagsByEntries || {};
    Object.keys(tagsByEntries).forEach(
      (key) =>
        (tagsByEntries[key] = castArray(tagsByEntries[key]).filter(Boolean)),
    );
    const shouldByEntries = Object.values(tagsByEntries).some(
      (entry) => Array.isArray(entry) && entry.length > 0,
    );

    // skip if options is empty.
    if (!tags.length && !shouldByEntries) {
      return;
    }

    const { HtmlTagsPlugin } = await import('../rspack/HtmlTagsPlugin');

    // create shared options used for entry without specified options.
    const sharedOptions: HtmlTagsPluginOptions = {
      append: true,
      hash: false,
      publicPath: true,
      tags,
    };
    // apply only one webpack plugin if `html.tagsByEntries` is empty.
    if (tags.length && !shouldByEntries) {
      chain
        .plugin(CHAIN_ID.PLUGIN.HTML_TAGS)
        .use(HtmlTagsPlugin, [sharedOptions]);
      return;
    }
    // apply webpack plugin for each entries.
    for (const [entry, filename] of Object.entries(api.getHTMLPaths())) {
      const opts = { ...sharedOptions, includes: [filename] };
      entry in tagsByEntries &&
        (opts.tags = (
          tagsByEntries as Record<string, HtmlInjectTagDescriptor[]>
        )[entry]);
      chain
        .plugin(`${CHAIN_ID.PLUGIN.HTML_TAGS}#${entry}`)
        .use(HtmlTagsPlugin, [opts]);
    }
  });
};

export const pluginHtml = (): RsbuildPlugin => ({
  name: 'rsbuild:html',

  setup(api) {
    api.modifyBundlerChain(
      async (chain, { HtmlPlugin, isProd, CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        // if html is disabled or target is server, skip html plugin
        if (isHtmlDisabled(config, target)) {
          return;
        }

        const minify = await getMinify(isProd, config);
        const assetPrefix = removeTailSlash(
          chain.output.get('publicPath') || '',
        );
        const entries = chain.entryPoints.entries() || {};
        const entryNames = Object.keys(entries);
        const htmlPaths = api.getHTMLPaths();
        const htmlInfoMap: Record<string, HtmlInfo> = {};

        await Promise.all(
          entryNames.map(async (entryName) => {
            const entryValue = entries[entryName].values() as string | string[];
            const chunks = getChunks(entryName, entryValue);
            const inject = getInject(entryName, config);
            const filename = htmlPaths[entryName];
            const { templatePath, templateContent } = await getTemplate(
              entryName,
              config,
              api.context.rootPath,
            );
            const templateParameters = getTemplateParameters(
              entryName,
              config,
              assetPrefix,
            );

            const metaTags = getMetaTags(entryName, config, templateContent);

            const pluginOptions: HTMLPluginOptions = {
              meta: metaTags,
              chunks,
              inject,
              minify,
              filename,
              template: templatePath,
              entryName,
              templateParameters,
              scriptLoading: config.html.scriptLoading,
            };

            const htmlInfo: HtmlInfo = {};
            htmlInfoMap[entryName] = htmlInfo;

            if (templateContent) {
              htmlInfo.templateContent = templateContent;
            }

            const title = getTitle(entryName, config);
            if (title) {
              pluginOptions.title = title;
            }

            const favicon = getFavicon(entryName, config);
            if (favicon) {
              if (isURL(favicon)) {
                htmlInfo.favicon = favicon;
              } else {
                // HTMLWebpackPlugin only support favicon file path
                pluginOptions.favicon = favicon;
              }
            }

            const finalOptions = mergeChainedOptions({
              defaults: pluginOptions,
              options: config.tools.htmlPlugin,
              utils: {
                entryName,
                entryValue,
              },
            });

            chain
              .plugin(`${CHAIN_ID.PLUGIN.HTML}-${entryName}`)
              .use(HtmlPlugin, [finalOptions]);
          }),
        );

        const { HtmlBasicPlugin } = await import('../rspack/HtmlBasicPlugin');

        chain
          .plugin(CHAIN_ID.PLUGIN.HTML_BASIC)
          .use(HtmlBasicPlugin, [{ info: htmlInfoMap }]);

        if (config.security) {
          const { nonce } = config.security;

          if (nonce) {
            const { HtmlNoncePlugin } = await import(
              '../rspack/HtmlNoncePlugin'
            );

            chain
              .plugin(CHAIN_ID.PLUGIN.HTML_NONCE)
              .use(HtmlNoncePlugin, [{ nonce }]);
          }
        }

        if (config.html) {
          const { appIcon, crossorigin } = config.html;

          if (crossorigin) {
            const { HtmlCrossOriginPlugin } = await import(
              '../rspack/HtmlCrossOriginPlugin'
            );

            const formattedCrossorigin =
              crossorigin === true ? 'anonymous' : crossorigin;

            chain
              .plugin(CHAIN_ID.PLUGIN.HTML_CROSS_ORIGIN)
              .use(HtmlCrossOriginPlugin, [
                { crossOrigin: formattedCrossorigin },
              ]);

            chain.output.crossOriginLoading(formattedCrossorigin);
          }

          if (appIcon) {
            const { HtmlAppIconPlugin } = await import(
              '../rspack/HtmlAppIconPlugin'
            );

            const distDir = getDistPath(config, 'image');
            const iconPath = path.isAbsolute(appIcon)
              ? appIcon
              : path.join(api.context.rootPath, appIcon);

            chain
              .plugin(CHAIN_ID.PLUGIN.APP_ICON)
              .use(HtmlAppIconPlugin, [{ iconPath, distDir }]);
          }
        }
      },
    );

    applyInjectTags(api);
  },
});
