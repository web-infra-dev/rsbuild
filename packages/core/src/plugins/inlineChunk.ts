import {
  pick,
  JS_REGEX,
  CSS_REGEX,
  isHtmlDisabled,
  type InlineChunkTest,
} from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

export const pluginInlineChunk = (): RsbuildPlugin => ({
  name: 'rsbuild:inline-chunk',

  setup(api) {
    api.modifyBundlerChain(async (chain, { target, CHAIN_ID, isDev }) => {
      const config = api.getNormalizedConfig();

      if (isHtmlDisabled(config, target) || isDev) {
        return;
      }

      const { InlineChunkHtmlPlugin } = await import(
        '../rspack/InlineChunkHtmlPlugin'
      );

      const { inlineStyles, inlineScripts } = config.output;

      const scriptTests: InlineChunkTest[] = [];
      const styleTests: InlineChunkTest[] = [];

      if (inlineScripts) {
        scriptTests.push(inlineScripts === true ? JS_REGEX : inlineScripts);
      }

      if (inlineStyles) {
        styleTests.push(inlineStyles === true ? CSS_REGEX : inlineStyles);
      }

      if (!scriptTests.length && !styleTests.length) {
        return;
      }

      chain
        .plugin(CHAIN_ID.PLUGIN.INLINE_HTML)
        // ensure nonce can be applied to inlined style tags
        .before(CHAIN_ID.PLUGIN.HTML_NONCE)
        .use(InlineChunkHtmlPlugin, [
          {
            styleTests,
            scriptTests,
            distPath: pick(config.output.distPath, ['js', 'css']),
          },
        ]);
    });
  },
});
