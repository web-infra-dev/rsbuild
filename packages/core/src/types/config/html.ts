import type { ConfigChainMergeContext, ConfigChainWithContext } from '../utils';
import type { OneOrMany } from '../utils';

export type CrossOrigin = 'anonymous' | 'use-credentials';

export type ScriptInject = boolean | 'body' | 'head';

export type ScriptLoading = 'defer' | 'module' | 'blocking';

export type OutputStructure = 'flat' | 'nested';

/**
 * custom properties
 * e.g. { name: 'viewport' content: 'width=500, initial-scale=1' }
 * */
export type MetaAttrs = { [attrName: string]: string | boolean };

export type MetaOptions = {
  /**
   * name content pair
   * e.g. { viewport: 'width=device-width, initial-scale=1, shrink-to-fit=no' }`
   * */
  [name: string]: string | false | MetaAttrs;
};

export type HtmlBasicTag = {
  tag: string;
  attrs?: Record<string, string | boolean | null | undefined>;
  children?: string;
};

export type HtmlTag = HtmlBasicTag & {
  hash?: boolean | string | ((url: string, hash: string) => string);
  publicPath?: boolean | string | ((url: string, publicPath: string) => string);
  append?: boolean;
  head?: boolean;
};

export type HtmlTagUtils = {
  hash: string;
  entryName: string;
  outputName: string;
  publicPath: string;
};

export type HtmlTagHandler = (
  tags: HtmlTag[],
  utils: HtmlTagUtils,
) => HtmlTag[] | void;

export type HtmlTagDescriptor = HtmlTag | HtmlTagHandler;

type ChainedHtmlOption<O> = ConfigChainMergeContext<O, { entryName: string }>;

export type AppIconItem = {
  /**
   * The path to the icon, can be a URL, an absolute file path,
   * or a relative path to the project root.
   */
  src: string;
  /**
   * The size of the icon.
   */
  size: number;
  /**
   * Specifies the intended target for which the icon should be generated.
   * - `apple-touch-icon` for iOS system.
   * - `web-app-manifest` for web application manifest.
   */
  target?: 'apple-touch-icon' | 'web-app-manifest';
};

export type AppIcon = {
  /**
   * The name of the application.
   * @see https://developer.mozilla.org/en-US/docs/Web/Manifest/name
   */
  name?: string;
  /**
   * The list of icons.
   */
  icons: AppIconItem[];
  /**
   * The filename of the manifest file.
   * @default 'manifest.webmanifest'
   */
  filename?: string;
};

export interface HtmlConfig {
  /**
   * Configure the `<meta>` tag of the HTML.
   */
  meta?: ChainedHtmlOption<MetaOptions>;
  /**
   * Set the title tag of the HTML page.
   */
  title?: ChainedHtmlOption<string>;
  /**
   * Set the inject position of the `<script>` tag.
   */
  inject?: ChainedHtmlOption<ScriptInject>;
  /**
   * Inject custom html tags into the output html files.
   */
  tags?: OneOrMany<HtmlTagDescriptor>;
  /**
   * Set the favicon icon for all pages.
   */
  favicon?: ChainedHtmlOption<string>;
  /**
   * Set the web application icons to display when added to the home screen of a mobile device.
   *
   * @example
   * appIcon: {
   *   name: 'My Website',
   *   icons: [
   *     { src: './icon-192.png', size: 192 },
   *     { src: './icon-512.png', size: 512 },
   *   ]
   * }
   */
  appIcon?: AppIcon;
  /**
   * Set the id of root element.
   */
  mountId?: string;
  /**
   * Set the [crossorigin](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/crossorigin) attribute
   * of the `<script>` tag.
   */
  crossorigin?: boolean | CrossOrigin;
  /**
   * Define the directory structure of the HTML output files.
   */
  outputStructure?: OutputStructure;
  /**
   * Define the path to the HTML template,
   * corresponding to the `template` config of [html-rspack-plugin](https://github.com/rspack-contrib/html-rspack-plugin).
   */
  template?: ChainedHtmlOption<string>;
  /**
   * Define the parameters in the HTML template,
   * corresponding to the `templateParameters` config of [html-rspack-plugin](https://github.com/rspack-contrib/html-rspack-plugin).
   */
  templateParameters?: ConfigChainWithContext<
    Record<string, unknown>,
    { entryName: string }
  >;
  /**
   * Set the loading mode of the `<script>` tag.
   */
  scriptLoading?: ScriptLoading;
}

export type NormalizedHtmlConfig = HtmlConfig & {
  meta: ChainedHtmlOption<MetaOptions>;
  title: ChainedHtmlOption<string>;
  mountId: string;
  inject: ChainedHtmlOption<ScriptInject>;
  crossorigin: boolean | CrossOrigin;
  outputStructure: OutputStructure;
  scriptLoading: ScriptLoading;
};
