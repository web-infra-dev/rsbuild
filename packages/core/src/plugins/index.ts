import {
  MEDIA_EXTENSIONS,
  FONT_EXTENSIONS,
  IMAGE_EXTENSIONS,
  Plugins,
} from '@rsbuild/shared';

export const plugins: Plugins = {
  html: () => import('./html').then((m) => m.pluginHtml()),
  cleanOutput: () => import('./cleanOutput').then((m) => m.pluginCleanOutput()),
  startUrl: () => import('./startUrl').then((m) => m.pluginStartUrl()),
  fileSize: () => import('./fileSize').then((m) => m.pluginFileSize()),
  devtool: () => import('./devtool').then((m) => m.pluginDevtool()),
  target: () => import('./target').then((m) => m.pluginTarget()),
  entry: () => import('./entry').then((m) => m.pluginEntry()),
  cache: () => import('./cache').then((m) => m.pluginCache()),
  yaml: () => import('./yaml').then((m) => m.pluginYaml()),
  toml: () => import('./toml').then((m) => m.pluginToml()),
  svg: () => import('./svg').then((m) => m.pluginSvg()),
  splitChunks: () => import('./splitChunks').then((m) => m.pluginSplitChunks()),
  inlineChunk: () => import('./inlineChunk').then((m) => m.pluginInlineChunk()),
  bundleAnalyzer: () =>
    import('./bundleAnalyzer').then((m) => m.pluginBundleAnalyzer()),
  font: () =>
    import('./asset').then((m) => m.pluginAsset('font', FONT_EXTENSIONS)),
  image: () =>
    import('./asset').then((m) => m.pluginAsset('image', IMAGE_EXTENSIONS)),
  media: () =>
    import('./asset').then((m) => m.pluginAsset('media', MEDIA_EXTENSIONS)),
  assetsRetry: () => import('./assetsRetry').then((m) => m.pluginAssetsRetry()),
  antd: () => import('./antd').then((m) => m.pluginAntd()),
  arco: () => import('./arco').then((m) => m.pluginArco()),
  checkSyntax: () => import('./checkSyntax').then((m) => m.pluginCheckSyntax()),
  rem: () => import('./rem').then((m) => m.pluginRem()),
  wasm: () => import('./wasm').then((m) => m.pluginWasm()),
  moment: () => import('./moment').then((m) => m.pluginMoment()),
  nodeAddons: () => import('./nodeAddons').then((m) => m.pluginNodeAddons()),
  externals: () => import('./externals').then((m) => m.pluginExternals()),
  networkPerformance: () =>
    import('./networkPerformance').then((m) => m.pluginNetworkPerformance()),
  preloadOrPrefetch: () =>
    import('./preloadOrPrefetch').then((m) => m.pluginPreloadOrPrefetch()),
  performance: () => import('./performance').then((m) => m.pluginPerformance()),
  define: () => import('./define').then((m) => m.pluginDefine()),
};
