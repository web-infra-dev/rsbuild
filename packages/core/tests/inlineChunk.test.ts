import { createStubRsbuild } from '@scripts/test-helper';
import { pluginEntry } from '../src/plugins/entry';
import { pluginHtml } from '../src/plugins/html';
import { pluginInlineChunk } from '../src/plugins/inlineChunk';

describe('plugin-inline-chunk', () => {
  beforeEach(() => {
    process.env.NODE_ENV = 'production';
  });

  afterEach(() => {
    process.env.NODE_ENV = 'test';
  });

  it('should use proper plugin options when inlineScripts is true', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginHtml(), pluginInlineChunk()],
      rsbuildConfig: {
        output: {
          inlineScripts: true,
        },
      },
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should use proper plugin options when inlineStyles is true', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginHtml(), pluginInlineChunk()],
      rsbuildConfig: {
        output: {
          inlineStyles: true,
        },
      },
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not apply InlineChunkHtmlPlugin when target is node', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginHtml(), pluginInlineChunk()],
      rsbuildConfig: {
        output: {
          target: 'node',
        },
      },
    });

    expect(
      await rsbuild.matchBundlerPlugin('InlineChunkHtmlPlugin'),
    ).toBeFalsy();
  });

  it('should not apply InlineChunkHtmlPlugin when target is web-worker', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginHtml(), pluginInlineChunk()],
      rsbuildConfig: {
        output: {
          target: 'web-worker',
        },
      },
    });

    expect(
      await rsbuild.matchBundlerPlugin('InlineChunkHtmlPlugin'),
    ).toBeFalsy();
  });

  it('should not apply InlineChunkHtmlPlugin in development mode', async () => {
    process.env.NODE_ENV = 'development';
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginHtml(), pluginInlineChunk()],
    });

    expect(
      await rsbuild.matchBundlerPlugin('InlineChunkHtmlPlugin'),
    ).toBeFalsy();
  });
});
