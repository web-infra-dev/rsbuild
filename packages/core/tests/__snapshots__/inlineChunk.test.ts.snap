// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-inline-chunk > should use proper plugin options when inlineScripts is true 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    InlineChunkHtmlPlugin {
      "distPath": {
        "css": "static/css",
        "js": "static/js",
      },
      "inlinedAssets": Set {},
      "name": "InlineChunkHtmlPlugin",
      "scriptTests": [
        /\\\\\\.\\(\\?:js\\|mjs\\|cjs\\|jsx\\)\\$/,
      ],
      "styleTests": [],
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-inline-chunk > should use proper plugin options when inlineStyles is true 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    InlineChunkHtmlPlugin {
      "distPath": {
        "css": "static/css",
        "js": "static/js",
      },
      "inlinedAssets": Set {},
      "name": "InlineChunkHtmlPlugin",
      "scriptTests": [],
      "styleTests": [
        /\\\\\\.css\\$/,
      ],
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;
